File: hilos.c
 /*
  *  Ejemplo de hilos desincronizados
  *
  *    Ing. Julian Perelli
  *    Catedra de Sistemas Operativos
  *    UTN - FRLP
  *    2012
  *
  * Posibles prerequisitos:
  *   apt-get install gcc build-essential glibc libc6
  *
  * Compilacion:
  *   gcc hilos.c -ohilos -lpthread
  *
  * Ejecucion:
  *   ./hilos
  *
  */

// libreria de funciones de hilos
#include <pthread.h>

// libreria de manejo de entrada/salida (stdin/out/err)
#include <stdio.h>

// directiva al compilador:
//   se reemplaza NUM_THREADS por el numero, antes de la fase de compilacion
#define NUM_THREADS     100


// funcion que ejecutara el hilo
void *PrintHello(void *threadid)
{
   // declaracion de variable long (mayor rango que int)
   long tid;

   // typecast (era void, ahora es long)
   tid = (long)threadid;

   // imprime en pantalla el thread id
   printf("Soy el hilo #%ld!\n", tid);

   // termina la ejecucion avisando al padre
   pthread_exit(NULL);
}


// funcion principal que se ejecuta
// argc es la cantidad de argumentos
// argv es un array de strings, son los argumentos
int main (int argc, char *argv[])
{
   // vector de threads (cada uno de tipo pthread)
   // longitud del vector = NUM_THREADS
   pthread_t threads[NUM_THREADS];

   // var donde se guarda el thread id
   long t;

   // bucle de repeticion for:
   //   desde t=0
   //   mientras t<NUM_THREADS
   //   realizando t++ (incremento de t en 1) en cada iteracion 
   for ( t = 0; t < NUM_THREADS; t++ ) {

      // Imprimir en pantalla un mensaje previo a la creacion de un hilo
      printf("En el main: creando hilo %ld\n", t);

      // Creacion del hilo:
      //   &threads[t]: puntero a thread id
      //   NULL: atributos de creacion (NULL=default)
      //   PrintHello: funcion que ejecutara el thread
      //   (void *)t: variable t (indice de este bucle), se pasa al thread, y la recibe como unico parametro
      pthread_create(&threads[t], NULL, PrintHello, (void *)t);

   }

   // lo ultimo que main debe hacer
   pthread_exit(NULL);
}

File: hilos.py
#!/usr/bin/python

 ##
  #  Ejemplo de hilos desincronizados
  #
  #    Ing. Julian Perelli
  #    Catedra de Sistemas Operativos
  #    UTN - FRLP
  #    2013
  #
  # Posibles prerequisitos:
  #   apt-get install python
  #
  # Ejecucio:
  #   python hilos.py
  #
  ##

# libreria de manejo de hilos
import threading

# libreria para enviar a dormir threads
from time import sleep

# clase que representa el hilo
class MiThread(threading.Thread):
    
    # funcion que se ejecuta al momento de crear el hilo
    ## self  : objeto de la clase MiThread
    ## numero: argumento pasado al momento de crear el hilo (idx)
    def __init__(self, numero):
        
        # imprimimos que el hilo esta siendo creado
        print "En el main: creando hilo " + str(i) + "\n",

        # llamamos a super init para la creacion de esta clase
        threading.Thread.__init__(self)
        
        # inicializamos una variable num para este hilo
        self.num = numero
    
    
    # funcion que se llama al ejecutar el hilo
    def run(self):
        
        # mandamos al hilo al estado sleep, aca se desincroniza
        sleep(1)
        
        # cuando vuelve de sleep, imprimimos el numero de hilo
        print "Soy el hilo " + str(self.num) + "\n",


## programa principal ##

# bucle de repeticion de 1 a 100
for idx in range(0,100):
    
    # creacion del hilo idx
    t = MiThread(idx)
    
    # ejecucion del hilo idx
    t.start()
